_format/1.1/index.md

```diff
@@ -1839,3 +1877,572 @@ An error object **MAY** have the following members:
+
+An operation object **MAY** contain any of the following members:
+
+* `op`: an operation code, expressed as a string, that indicates the type of
+  operation to perform. Valid values include:
+  - `"get"`: fetches a resource, resource identity, collection of resources, or
+     collection of resource identitities
+  - `"add"`: creates a new resource or relationship
+  - `"update"`: updates a resource or relationship
+  - `"remove"`: removes a resource or relationship
+
+* `data`: the operation's "primary data".
+
+* `included`: an array of [resource
+  objects](#document-resource-objects) that are
+  related to the operation's primary data and/or each other
+  ("included resources").
+
+* `ref`: the resource or resource collection that is the target of the
+  operation. This member serves to disambiguate the target at a particular
+  endpoint if that endpoint represents multiple resources. `ref` **MUST** be
+  expressed as an object that **MAY** contain any of the following members:
+  - `type`: resource `type`
+  - `id` or `lid`: resource identifier or local identifier
+  - `relationship`: the name of the relationship
+
+* `params`: all of the query parameters allowed for singular requests can be
+  represented in JSON form in an operation's `params` object. Possible members
+  include:
+  - `include`
+  - `sort`
+  - `filter`
+  - `page`
+  - `fields`
+
+* `links`: a [links object](#document-links) related to the operation's primary
+  data.
+
+* `meta`: a [meta object](#document-meta) that contains non-standard
+  meta-information about the operation.
+
+Different members are required for processing different types of operations, as
+described below.
+
+> Note: `op` codes are used instead of the HTTP methods prescribed for singular
+  requests. The values `"get"`, `"add"`, `"update"`, and `"remove"` align with
+  the HTTP methods `GET`, `POST`, `PATCH`, and `DELETE` used for singular
+  requests.
+
+### <a href="#processing-operations" id="processing-operations" class="headerlink"></a> Processing Operations
+
+A server **MUST** perform operations in the order they appear in the
+`operations` array.
+
+All operations in a request **MUST** be performed atomically, so that a failure
+to perform any operation **MUST** invalidate any effects of preceding
+operations.
+
+A successful operation's `data` and `included` members **MUST** conform with the
+requirements for the analogous singular request and response.
+
+A server **MUST** respond to successful operations requests with `200 OK` if
+any operations are required to have `data` in a response. An array of resultant
+operation objects **MUST** be returned that correspond positionally to the
+requested operations.
+
+A server **MAY** respond to successful requests that include operations with
+`204 No Content` and no response document if no operations are required to
+return `data`.
+
+If any operation in a request fails, the server **MUST** respond as described
```

Comment:
What I'm missing here, is the notion of *partial success*.

A single _operations request_ could:
1. create a resource (f.e. create a comment that is a reply to another comment)
2. fetch resources (f.e. fetch the next page of comments)
3. …

Any of those could fail, and in different degrees:
1. the comment I'm posting could be a reply to a comment that has since been deleted … but I'd still want the other operations to succeed
2. some of those comments I may not be able to access because they're still in moderation
3. …

In either of those cases, I would expect to see the things that did succeed to actually return the expected data … but this pointing to `[Errors]` (http://jsonapi.org/format/#errors) means that is not possible, because as http://jsonapi.org/format/#document-top-level says:

> The members `data` and `errors` *MUST NOT* coexist in the same document.

---

I suspect the answer is going to be along the lines of "you should use operations requests not for bulk operations, but for transactions, i.e. for operations that should atomically succeed or fail". @dgeb stated something along those lines in https://github.com/json-api/json-api/pull/1254#issuecomment-364966578.

---

Which then leaves me wondering if we're not misnaming this PR and the concepts being added to the spec. Because @EndangeredMassa wrote specifically in https://github.com/json-api/json-api/pull/1254#issuecomment-365380315 that "bulk operations hinders adoptability". I suspect many of us see "bulk operations" and conclude that this PR brings that.

He even asked specifically:
> That tooling could be simplified if it always used operations, right?

Which @dgeb answered with:
> Yes, tooling could use operations exclusively without losing any functionality from the current spec.

But that's not entirely accurate as far as I can tell. *Merging many singular requests into a single operations request _does_ lose functionality*: any of those singular requests can result in errors without affecting the other singular requests. But if those singular requests are merged into a single operations request, then any single operation failing requires the entire operations request to fail.
Which means that tooling could *not* exclusively use operations.

---

*I'm wondering whether this should remove all mentions of "bulk" and "operations", and instead use "transactions" in both the spec and as the title of this PR.*

Quoting @dgeb from https://github.com/json-api/json-api/pull/1254#issuecomment-365291490:
>  The only additional requirements are accounting for local IDs and transactionality

Reply:
Am I summarizing this correctly @wimleers ? You're suggesting:

The PR and new top-level member, `operations`, should be renamed to better reflect that this proposal is about _atomicity_, not bulk processing.

Bulk processing, i.e. bundling requests, is actually an anti-pattern where atomicity is not required.

See also, @dgeb in https://github.com/json-api/json-api/pull/1254#issuecomment-364932213
> If order is not important, then I would strongly recommend making multiple singular requests in parallel.

Reply:
The transactional nature of this proposal is actually consistent with the [current spec](http://jsonapi.org/format/#crud): 

> A request MUST completely succeed or fail (in a single “transaction”). No partial updates are allowed.

This prevents changes to one attribute or relationship from succeeding while others fail. The operations proposal just extends this further.

Reply:
> But that's not entirely accurate as far as I can tell. Merging many singular requests into a single operations request does lose functionality: any of those singular requests can result in errors without affecting the other singular requests. But if those singular requests are merged into a single operations request, then any single operation failing requires the entire operations request to fail.
> Which means that tooling could not exclusively use operations.

What I meant here was that the tooling could exclusively use operations if it always used one operation per request, not that it could necessarily submit multiple operations all the time.

---

Requiring this be transactional means that it's serving some uses well, but others not as well. You listed out use cases up top, but I think there are more to consider. I named yours for reference:

> **Compound Document:** Creating multiple resources of the same or different types together and then creating relationships between them. For example, creating an invoice together with one or more invoice items.

Transactionality here seems necessary. If any part of this compound resource fails, you likely want the entire thing to fail.

> **Add/Remove Relationships:** Editing memberships of very large relationships, assuming that edits include additions and deletions. For example, removing two members of an organization at the same time two other members are added.

Transactionality here seems optional. If I've done a lot of relationship remapping and one specific change fails, then I definitely want to know about that. However, I don't think I necessarily want them all to fail.

> **Cascading Deletes:** Clearing relationships and explicitly deleting the related resources as well. For example, deleting an author and some of the articles associated with that author.

Transactionality here seems necessary. I think we'd suggest to API developers to cascade deletes on the server side responsibly, but if an API doesn't do that, you'd want to submit a series of operations like this in a transactional manner.

> **Offline Sync:** Sending a general "batch" of changes that must succeed or fail together. For example, queuing up an array of operations while offline and then applying them together.

Transactionality here seems problematic. It means our offline sync is more likely to fail entirely. I think we'd rather have partial success and show the individual failures to the user. Should a change conflict on a single record fail your entire offline sync that creates/updates/deletes many different records?

**New Use Case: Network Performance Sensitivity**: A latency-sensitive application optimizing for every drop of network performance wants to submit combine as many operations as it can, whenever it can. (Example: A mobile application running in a poor-connection area wants to update an auto-buy stock record AND get the latest stock prices for your tracked stocks.)

Transactionality here means seems problematic. It means that your unrelated operations can fail each other, which will require follow-up requests, possibly splitting them up into multiple requests to avoid the transactionality of the system. Should my resource update fail my unrelated resource read?

---

The use cases around related records seem to benefit from transactionality, but the use cases around unrelated records seems to be limited by it. I think that this shows that transactionality should be optional in some way. In my opinion, if transactionality is required, bulk operations won't be as useful.

Reply:
@EndangeredMassa thanks for the thorough write up!

My understanding of the proposal thus far is that it's only intended to permit transactionality/atomicity and that the proposal wasn't intending to enable "bulk" operations like your final "network performance" example. While it would be _possible_, it wouldn't be recommended.

The commit summary which introduced this reads:

> Operations can be used to perform one or more fetch or mutation operations serially and transactionally.

I think this is exactly why @wimleers said:

> I'm wondering whether this should remove all mentions of "bulk" and "operations", and instead use "transactions" in both the spec and as the title of this PR.

Reducing the perceived scope of this proposal _just_ to transactionality with a few language changes will guide users to its intended usage. Like so:

~Introduce Operations to v1.1~ Introduce Atomic Operations to v1.1

```diff
 A server **MAY** support usage of the `operations` array to complete one or more
-fetch or mutation operations in a single request.
+fetch or mutation operations in a single transaction.
...
-  PATCH /bulk HTTP/1.1
+  PATCH /transaction HTTP/1.1
```

Reply:
Does the scope creep too much if we remove that requirement? If so, I can understand keeping it for now and proposing a separate thing that could build off of the work done in this proposal. I would then also support making it more explicit that this is for transactions.

Reply:
I think it's impossible to remove the transactional requirement in this proposal:

> A request MUST completely succeed or fail (in a single “transaction”). No partial updates are allowed.  — http://jsonapi.org/format/#crud

Bulk operations with partial success would be a hard break from v1 of the spec.

Which would mean @gabesullice (me), @wimleers, and you (@EndangeredMassa) all "support making it more explicit that this is for transactions."

@dgeb, what are your thoughts? Have we missed anything?

Reply:
> Transactionality here seems problematic. It means our offline sync is more likely to fail entirely. I think we'd rather have partial success and show the individual failures to the user.

@EndangeredMassa I agree with your analysis here. The way I presented this use case was a bit sloppy, because it's not how I would handle offline sync either. I would send operations together that map to a logical user action (say, saving a form) in a single request. And I would serially send requests until the offline queue was cleared, so that client code could reason about failure cases.

There is another level this proposal could go, in which we allow an array of operations to be grouped together into a transaction, and we send an array of transactions. However, this could be handled at a later date in a separate non-breaking proposal that relies on a top-level `transactions` array. Something like this may seem ideal approach for offline sync because it could be quite efficient, but there are also a lot of non-obvious decisions re: responses that represent mixed results. I'd rather not muddy the waters of the operations proposal at this time by expanding the scope this much.

Reply:
> Bulk operations with partial success would be a hard break from v1 of the spec.

@gabesullice I think we could technically create new rules for operations requests distinct from the rules for singular requests, since operations are a new feature. However, I am not in favor of introducing non-atomic requests now (see my response to @EndangeredMassa).